### PR TITLE
feat(partitioning): Add partition specs/fields and basic transform interface

### PR DESCRIPTION
**Pull request by zeroshade**
Adds representations for Partition Specs, Partition Fields, and base transform interface + implementations. It does not add actual transformation functionality implementations yet. That will come later when implementing data reading. This adds only what is sufficient for reading a table definition / metadata.

### CODE REVIEW

File: partitions.go [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+// UnpartitionedPartitionSpec is the default unpartitioned spec which can
+// be used for comparisons or to just provide a convenience for referencing
+// the same unpartitioned spec object.
+var UnpartitionedPartitionSpec = &PartitionSpec{id: 0}
+
+// PartitionField represents how one partition value is derived from the
+// source column by transformation.
+type PartitionField struct {
+	// SourceID is the source column ID of the table's schema
```
Comment (coded9) [inline partitions.go]: nit: column id instead of column ID in the comment to maintain consistency

File: partitions.go [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+// data reading gets implemented.
+type Transform interface {
+	fmt.Stringer
+	encoding.TextMarshaler
+	ResultType(t Type) Type
+}
+
+// IdentityTransform uses the identity function, performing no transformation
+// but instead partitioning on the value itself.
+type IdentityTransform struct{}
```
Comment (coded9) [inline partitions.go]: Any reason why _transforms_ code is clubbed with _partitioning_ ?

Comment (zeroshade) [reply partitions.go]: mostly because the `PartitionField` has a `Transform` object inside of it. Anything else that interacts with transforms, at least currently, only does so by first going through the partition spec and partition fields. I'm fine with moving the Transform code to it's own `transform.go` file if you think that would make more sense

File: partitions.go (hunk 1 above)
Comment (zeroshade) [reply partitions.go]: done

File: partitions.go (hunk 2 above)
Comment (Fokko) [reply partitions.go]: I think splitting the `transforms` into a separate `transform.go` makes sense

File: partitions_test.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/apache/iceberg-go"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseTransform(t *testing.T) {
```
Comment (Fokko) [inline partitions_test.go]: In Java we ignore the casing: https://github.com/apache/iceberg/commit/e736d6dfbac16bef6f5f7ab4fe73bd9a27fe3f82 I think we should do that here as well

File: partitions_test.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			require.NoError(t, err)
+			assert.Equal(t, tt.expected, transform)
+
+			txt, err := transform.MarshalText()
+			assert.NoError(t, err)
+			assert.Equal(t, tt.toparse, string(txt))
+		})
+	}
+
+	errorTests := []struct {
```
Comment (Fokko) [inline partitions_test.go]: 👍🏻

File: partitions_test.go (hunk 3 above)
Comment (zeroshade) [reply partitions_test.go]: Updated to ignore case (and updated test cases)

File: partitions.go (hunk 2 above)
Comment (zeroshade) [reply partitions.go]: updated and added.

**Comment by coded9** [comment]
> @nastra @coded9 @Fokko @rdblue @bitsondatadev Any further comments / reviews here?

LGTM!

**Comment by zeroshade** [comment]
@nastra @rdblue any chance for a review soon to get this in?

File: partitions.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+// under the License.
+
+package iceberg
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/slices"
```
Comment (JustinTimperio) [inline partitions.go:25]: Generally I am really against using the experimental packages for anything that will be used likely in production. There recently has been a lot of people ignoring this and creating havoc in the release cycles of Go as seen in the following issues:

https://github.com/golang/go/issues/51317
https://github.com/golang/go/issues/61374

I think it would be better at this time to avoid using anything in the exp branch.

File: partitions.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+func (ps *PartitionSpec) CompatibleWith(other *PartitionSpec) bool {
+	if ps == other {
+		return true
+	}
+
+	if len(ps.fields) != len(other.fields) {
+		return false
+	}
+
+	return slices.EqualFunc(ps.fields, other.fields, func(left, right PartitionField) bool {
```
Comment (JustinTimperio) [inline partitions.go:108]: Happy to write a replacement function for this to avoid the exp package

File: transforms.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+func (VoidTransform) String() string { return "void" }
+
+func (VoidTransform) ResultType(t Type) Type { return t }
+
+// BucketTransform transforms values into a bucket partition value. It is
+// parameterized by a number of buckets. Bucket partition transforms use
+// a 32-bit hash of the source value to produce a positive value by mod
+// the bucket number.
+type BucketTransform struct {
+	N int
```
Comment (nastra) [inline transforms.go]: maybe rename to `numBuckets`?

File: transforms.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+	return []byte(t.String()), nil
+}
+
+func (t BucketTransform) String() string { return fmt.Sprintf("bucket[%d]", t.N) }
+
+func (BucketTransform) ResultType(Type) Type { return PrimitiveTypes.Int32 }
+
+// TruncateTransform is a transformation for truncating a value to a specified width.
+type TruncateTransform struct {
+	W int
```
Comment (nastra) [inline transforms.go]: maybe rename to `width`?

File: transforms.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+		case "hour":
+			return HourTransform{}, nil
+		}
+	}
+
+	return nil, fmt.Errorf("%w: %s", ErrInvalidTransform, s)
+}
+
+// Transform is an interface for the various Transformation types
+// in partition specs. Currently they do not yet provide actual
```
Comment (nastra) [inline transforms.go]:
```suggestion
// in partition specs. Currently, they do not yet provide actual
```

File: partitions_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+func TestPartitionSpec(t *testing.T) {
+	assert.Equal(t, 1000, iceberg.UnpartitionedPartitionSpec.LastAssignedFieldID())
+
+	bucket := iceberg.BucketTransform{N: 4}
+	idField1 := iceberg.PartitionField{
+		SourceID: 3, FieldID: 1001, Name: "id", Transform: bucket}
+	spec1 := iceberg.NewPartitionSpec(idField1)
+
+	assert.Zero(t, spec1.ID())
```
Comment (nastra) [inline partitions_test.go:37]: independent from this PR, is there a good fluent assertion library available in the Golang ecosystem (similar to AssertJ)?

File: partitions.go [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/slices"
+)
+
+const (
+	partitionDataIDStart   = 1000
```
Comment (nastra) [inline partitions.go:29]: minor: any reason why this one is lowercase while the other constant is uppercase?

File: partitions_test.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+	"encoding/json"
+	"testing"
+
+	"github.com/apache/iceberg-go"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPartitionSpec(t *testing.T) {
+	assert.Equal(t, 1000, iceberg.UnpartitionedPartitionSpec.LastAssignedFieldID())
```
Comment (nastra) [inline partitions_test.go]: I believe in java the last assigned field ID for an unpartitioned spec returns 999: https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L320

File: partitions.go [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+	InitialPartitionSpecID = 0
+)
+
+// UnpartitionedPartitionSpec is the default unpartitioned spec which can
+// be used for comparisons or to just provide a convenience for referencing
+// the same unpartitioned spec object.
+var UnpartitionedPartitionSpec = &PartitionSpec{id: 0}
```
Comment (nastra) [inline partitions.go]: minor: maybe just `UnpartitionedSpec`?

File: partitions.go [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+	for _, f := range ps.fields {
+		ps.sourceIdToFields[f.SourceID] =
+			append(ps.sourceIdToFields[f.SourceID], f)
+	}
+}
+
+func (ps *PartitionSpec) ID() int                    { return ps.id }
+func (ps *PartitionSpec) NumFields() int             { return len(ps.fields) }
+func (ps *PartitionSpec) Field(i int) PartitionField { return ps.fields[i] }
+func (ps *PartitionSpec) IsUnpartitioned() bool      { return len(ps.fields) == 0 }
```
Comment (nastra) [inline partitions.go]: do we need to consider void transforms, like we do in the Java impl?

File: partitions.go (hunk 11 above)
Comment (zeroshade) [reply partitions.go:29]: I wasn't exporting this one (for Go the first letter being uppercase or lowercase determines whether it is exported) while I am exporting the other constant.

File: partitions_test.go (hunk 10 above)
Comment (zeroshade) [reply partitions_test.go:37]: I'm not familiar with AssertJ, but the most common assertion library I've seen for testing in the Golang ecosystem is the testify library that I'm using here.

File: partitions_test.go (hunk 12 above)
Comment (zeroshade) [reply partitions_test.go]: fixed

File: partitions.go (hunk 14 above)
Comment (zeroshade) [reply partitions.go]: good catch, i didn't realize that. Fixed

File: partitions.go (hunk 5 above)
Comment (nastra) [reply partitions.go:25]: @zeroshade thoughts on this one?

File: partitions.go [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+		ps.sourceIdToFields[f.SourceID] =
+			append(ps.sourceIdToFields[f.SourceID], f)
+	}
+}
+
+func (ps *PartitionSpec) ID() int                    { return ps.id }
+func (ps *PartitionSpec) NumFields() int             { return len(ps.fields) }
+func (ps *PartitionSpec) Field(i int) PartitionField { return ps.fields[i] }
+
+func (ps *PartitionSpec) IsUnpartitioned() bool {
```
Comment (nastra) [inline partitions.go:157]: just to make sure: this will return true only if 1) fields are empty or 2) all transforms are void

File: partitions_test.go [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+	assert.True(t, spec1.CompatibleWith(&spec2))
+	assert.Equal(t, []iceberg.PartitionField{idField1}, spec1.FieldsBySourceID(3))
+	assert.Empty(t, spec1.FieldsBySourceID(1925))
+
+	spec3 := iceberg.NewPartitionSpec(idField1, idField2)
+	assert.False(t, spec1.CompatibleWith(&spec3))
+	assert.Equal(t, 1002, spec3.LastAssignedFieldID())
+}
+
+func TestUnpartitionedWithVoidField(t *testing.T) {
```
Comment (nastra) [inline partitions_test.go:62]: maybe also add a test where one field has a void transform and another field has some non-void transform, thus resulting `spec.IsUnpartitioned()` to be false

**Review by nastra — APPROVED** [review]
overall LGTM, just left one more comment around testing

File: partitions.go (hunk 5 above)
Comment (zeroshade) [reply partitions.go:25]: Sorry I forgot to respond here on the PR because we ended up discussing this in the slack channel. 

Tl;DR: the standard for Go in general is to target N-1 versions (ie. right now the latest is Go 1.21, so we want to target Go 1.20 + Go 1.21). This slices package was only added to the Go stdlib in Go 1.21, so we still need to point at `golang.org/x/exp/slices` until/unless we drop support for Go 1.20, at which point we can update this to just use the stdlib version of the package. 

We're still about 6 months or so before Go 1.22 would get released, but this should be fine for now. Once Go 1.22 is released I'll replace these with the stdlib package, letting us maintain the N-1 compatibility.

File: partitions.go (hunk 15 above)
Comment (zeroshade) [reply partitions.go:157]: Correct, I've added tests accordingly.

File: partitions_test.go (hunk 16 above)
Comment (zeroshade) [reply partitions_test.go:62]: Added the test

**Comment by nastra** [comment]
thanks for working on this @zeroshade! Should be good to go once the merge conflict is resolved

**Comment by zeroshade** [comment]
@nastra resolved! :smile:
